src/audio/multiband_drc/multiband_drc_generic.c

```diff
@@ -0,0 +1,411 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <stdint.h>
+#include <sof/audio/drc/drc_helper.h>
+#include <sof/audio/format.h>
+#include <sof/audio/eq_iir/iir.h>
+#include <sof/audio/multiband_drc/multiband_drc.h>
+
+#if CONFIG_FORMAT_S16LE
+static void multiband_drc_s16_default_pass(const struct comp_dev *dev,
+					   const struct audio_stream *source,
+					   struct audio_stream *sink,
+					   uint32_t frames)
+{
+	int16_t *x;
+	int16_t *y;
+	int i;
+	int n = source->channels * frames;
+
+	for (i = 0; i < n; i++) {
+		x = audio_stream_read_frag_s16(source, i);
+		y = audio_stream_write_frag_s16(sink, i);
+		*y = *x;
+	}
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE
+static void multiband_drc_s32_default_pass(const struct comp_dev *dev,
+					   const struct audio_stream *source,
+					   struct audio_stream *sink,
+					   uint32_t frames)
+{
+	int32_t *x;
+	int32_t *y;
+	int i;
+	int n = source->channels * frames;
+
+	for (i = 0; i < n; i++) {
+		x = audio_stream_read_frag_s32(source, i);
+		y = audio_stream_write_frag_s32(sink, i);
+		*y = *x;
+	}
+}
+#endif /* CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE */
+
+static void multiband_drc_process_emp_crossover(struct multiband_drc_state *state,
+						crossover_split split_func,
+						int32_t *buf_src,
+						int32_t *buf_sink,
+						int enable_emp,
+						int nch,
+						int nband)
+{
+	struct iir_state_df2t *emp_s;
+	struct crossover_state *crossover_s;
+	int32_t *buf_sink_band;
+	int ch, band;
+	int32_t emp_out;
+	int32_t crossover_out[nband];
+
+	for (ch = 0; ch < nch; ch++) {
+		emp_s = &state->emphasis[ch];
+		crossover_s = &state->crossover[ch];
+
+		if (enable_emp)
+			emp_out = iir_df2t_2stage_biquads(emp_s, *buf_src);
+		else
+			emp_out = *buf_src;
+
+		split_func(emp_out, crossover_out, crossover_s);
+		buf_sink_band = buf_sink;
+		for (band = 0; band < nband; band++) {
+			*buf_sink_band = crossover_out[band];
+			buf_sink_band += PLATFORM_MAX_CHANNELS;
+		}
+
+		buf_src++;
+		buf_sink++;
+	}
+}
+
+static void multiband_drc_process_drc(struct drc_state *state,
+				      const struct sof_drc_params *p,
+				      int32_t *buf_src,
+				      int32_t *buf_sink,
+				      int nbyte,
+				      int nch)
+{
+	int16_t *pd_write16;
+	int16_t *pd_read16;
+	int32_t *pd_write32;
+	int32_t *pd_read32;
+	int ch;
+	int pd_write_index;
+	int pd_read_index;
+	int is_2byte = (nbyte == 2); /* otherwise is 4-bytes */
+
+	if (p->enabled && !state->processed) {
+		drc_update_envelope(state, p);
+		drc_compress_output(state, p, nbyte, nch);
+		state->processed = 1;
+	}
+
+	pd_write_index = state->pre_delay_write_index;
+	pd_read_index = state->pre_delay_read_index;
+
+	if (is_2byte) { /* 2 bytes per sample */
```

Comment:
hmm, im not the biggest fan of this check, but im not sure how to fix this other than splitting the function

Reply:
ack - and doing so makes the conversion to HiFi intrinsics easier to follow.

Reply:
@johnylin76 friendly ping on this